README.md
# react-native-auto-height-image

Initialized by [vivaxy/gt-npm-package](https://github.com/vivaxy/gt-npm-package)

[![NPM Version](http://img.shields.io/npm/v/react-native-auto-height-image.svg?style=flat-square)](https://www.npmjs.com/package/react-native-auto-height-image)
[![NPM Downloads](https://img.shields.io/npm/dt/react-native-auto-height-image.svg?style=flat-square)](https://www.npmjs.com/package/react-native-auto-height-image)
[![MIT License](https://img.shields.io/npm/l/react-native-auto-height-image.svg?style=flat-square)](./LICENSE)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg?style=flat-square)](https://conventionalcommits.org)
[![Financial Contributors on Open Collective](https://opencollective.com/react-native-auto-height-image/all/badge.svg?label=financial+contributors&style=flat-square)](https://opencollective.com/react-native-auto-height-image)
[![Maintainers Wanted](https://img.shields.io/badge/maintainers-wanted-red.svg?style=flat-square)](https://github.com/vivaxy/react-native-auto-height-image/issues/88)

This component provides you a simple way to load a remote image and automatically set `Image` height to the image dimension which fits the provided width.

React Native `Image` component needs users to set both `width` and `height` props.

React Native version requirements: >=0.46.

## Installation

`yarn add react-native-auto-height-image`

`npm install react-native-auto-height-image`

## Usage

Use local or remote files:

```js
import React, { Component } from 'react';
import AutoHeightImage from 'react-native-auto-height-image';

import image from 'gallifrey-falls.png';

export default class Demo extends Component {
  render() {
    return (
      <View>

        <AutoHeightImage
          width={100}
          source={image}
        />

        <AutoHeightImage
          width={100}
          source={{uri: 'http://placehold.it/350x150'}}
        />

      </View>
    );
  }
}
```

You can even specify fallback images for when the source fails to load:

```js
import React, { Component } from 'react';
import AutoHeightImage from 'react-native-auto-height-image';

import image from 'gallifrey-falls.png';

export default class Demo extends Component {
  render() {
    return (
      <AutoHeightImage
        width={100}
        source={{uri: 'https://vivaxy.github.io/404'}}
        fallbackSource={image}
      />
    );
  }
}
```

### Props

| name               | type             | isRequired    | default           | description                                                           |
| ---                | ---              | ---           | ---               | ---                                                                   |
| `width`            | number           | ✔             | N/A               | image width to fit                                                    |
| `maxHeight`            | number           | ✖             | `Infinity`               | image max height                                                    |
| `source`           | number or object | ✔             | N/A               | local (i.e. require/import) or remote image ({uri: '...'})            |
| `fallbackSource`   | number or object | ✖             | N/A               | local (i.e. require/import) or remote image ({uri: '...'})            |
| `onHeightChange`   | func             | ✖             | `(height) => {}`    | called when updating image height, the argument `height` might be `0` |
| `animated`        | bool              | ✖              | `false`               | Use `Animated.Image` instead of `Image` |

Other [image props](https://reactnative.dev/docs/image#props) except `resizeMode` are accepted.

## Change Log

[Change log](./CHANGELOG.md)

## Contributing

[Contributing](./CONTRIBUTING.md)

## Licence

[MIT](./LICENSE)

## Contributors

### Code Contributors

This project exists thanks to all the people who contribute. [[Contribute](CONTRIBUTING.md)].
<a href="https://github.com/vivaxy/react-native-auto-height-image/graphs/contributors"><img src="https://opencollective.com/react-native-auto-height-image/contributors.svg?width=890&button=false" /></a>

### Financial Contributors

Become a financial contributor and help us sustain our community. [[Contribute](https://opencollective.com/react-native-auto-height-image/contribute)]

#### Individuals

<a href="https://opencollective.com/react-native-auto-height-image"><img src="https://opencollective.com/react-native-auto-height-image/individuals.svg?width=890"></a>

#### Organizations

Support this project with your organization. Your logo will show up here with a link to your website. [[Contribute](https://opencollective.com/react-native-auto-height-image/contribute)]

<a href="https://opencollective.com/react-native-auto-height-image"><img src="https://opencollective.com/react-native-auto-height-image/organization.svg?width=890"></a>

## Related Projects

- [react-native-scalable-image](https://github.com/ihor/react-native-scalable-image)
- [react-native-fit-image](https://github.com/huiseoul/react-native-fit-image)
- [react-native-responsive-image-view](https://github.com/wKovacs64/react-native-responsive-image-view)
- [react-native-auto-image](https://github.com/egorshulga/react-native-auto-image)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.2.4](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.2.3...v3.2.4) (2021-02-04)


### Features

* supports `ImageBackground` [@SoniaComp](https://github.com/SoniaComp)

### [3.2.3](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.2.2...v3.2.3) (2020-10-17)


### Bug Fixes

* **deps:** update dependency expo to v39 ([9cf60e4](https://github.com/vivaxy/react-native-auto-height-image/commit/9cf60e4798742a4ef648cc39cd96af6380b7180e))
* **deps:** update dependency react to v16.14.0 ([ed45e7e](https://github.com/vivaxy/react-native-auto-height-image/commit/ed45e7ebf50b677ccdc501a01ae813258660cfc6))
* **deps:** update dependency react-native-web to ^0.14.0 ([1ad617f](https://github.com/vivaxy/react-native-auto-height-image/commit/1ad617fa514cb307fc72a3c34e8254843e19d9d0))

### [3.2.2](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.2.1...v3.2.2) (2020-07-25)


### Bug Fixes

* **deps:** update dependency expo to v38 ([6b24bd8](https://github.com/vivaxy/react-native-auto-height-image/commit/6b24bd8c1a2f1bf7fafcb0ba3c7b558318d7d152))
* **deps:** update dependency react-native-web to ^0.13.0 ([82d6270](https://github.com/vivaxy/react-native-auto-height-image/commit/82d6270e7e45855864fcc4fcf1ec62ee97695300))

### [3.2.1](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.2.0...v3.2.1) (2020-06-08)

## [3.2.0](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.1.3...v3.2.0) (2020-05-27)


### Features

* :sparkles: support `maxHeight` prop, reuse image cache on initial rendering to optimize the performance ([e5055c8](https://github.com/vivaxy/react-native-auto-height-image/commit/e5055c8e97a800581d7049140392cdedad035f48))

### [3.1.3](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.1.2...v3.1.3) (2020-05-07)


### Bug Fixes

* :bug: fix `Can't perform a React state update on an unmounted component.` ([035e6a8](https://github.com/vivaxy/react-native-auto-height-image/commit/035e6a86ba39daf3c16d75d7467f925d94537023)), closes [#44](https://github.com/vivaxy/react-native-auto-height-image/issues/44)

### [3.1.2](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.1.1...v3.1.2) (2020-04-18)


### Bug Fixes

* add missing symbol ([7498b82](https://github.com/vivaxy/react-native-auto-height-image/commit/7498b823bb8ac8a0f1f509fb0f083a61e57e192d))

### [3.1.1](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.1.0...v3.1.1) (2020-04-04)


### Bug Fixes

* android crash at imagePolyfill ([21dc084](https://github.com/vivaxy/react-native-auto-height-image/commit/21dc0841c452a4178202db1beedfc7e2e72d7665))

## [3.1.0](https://github.com/vivaxy/react-native-auto-height-image/compare/v3.0.0...v3.1.0) (2020-03-31)


### Features

* animated image support ([97ff786](https://github.com/vivaxy/react-native-auto-height-image/commit/97ff786c40143599228524c1b12d4e29ba496e57))

## [3.0.0](https://github.com/vivaxy/react-native-auto-height-image/compare/v2.0.0...v3.0.0) (2020-01-17)


### ⚠ BREAKING CHANGES

* **changelog:** Drop support for react before 16.8 and react-native before 0.59

* **changelog:** :memo: commit a breaking change ([03c9b81](https://github.com/vivaxy/react-native-auto-height-image/commit/03c9b81))


### Bug Fixes

* fix for fallback images ([d825375](https://github.com/vivaxy/react-native-auto-height-image/commit/d825375))


### Features

* **deps:** bumped React to ^16.8 & React Native to ^0.59.0 ([24edbc2](https://github.com/vivaxy/react-native-auto-height-image/commit/24edbc2))

## [2.0.0](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.1.3...v2.0.0) (2020-01-09)


### ⚠ BREAKING CHANGES

* incorporated imagepolyfill in codebase, removed from deps

### Features

* incorporated imagepolyfill in codebase, removed from deps ([bfdca6f](https://github.com/vivaxy/react-native-auto-height-image/commit/bfdca6f))

<a name="1.1.3"></a>
## [1.1.3](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.1.2...v1.1.3) (2019-10-16)



<a name="1.1.1"></a>
## [1.1.1](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.1.0...v1.1.1) (2019-08-13)


### Reverts

* **open collective:** :rewind: revert changes to package.json ([37f3b17](https://github.com/vivaxy/react-native-auto-height-image/commit/37f3b17))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.0.5...v1.1.0) (2019-03-07)


### Features

* :sparkles:updateImageHeight with safe check ([a2c9275](https://github.com/vivaxy/react-native-auto-height-image/commit/a2c9275))



<a name="1.0.5"></a>
## [1.0.5](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.0.4...v1.0.5) (2018-10-22)


### Bug Fixes

* :bug:Fix updating the image size after change of the source ([cfe1566](https://github.com/vivaxy/react-native-auto-height-image/commit/cfe1566))



<a name="1.0.4"></a>
## [1.0.4](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.0.3...v1.0.4) (2018-10-09)


### Bug Fixes

* :bug:Remove trailing comma ([89713a5](https://github.com/vivaxy/react-native-auto-height-image/commit/89713a5))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.0.2...v1.0.3) (2018-10-09)


### Features

* Allowing Image props.



<a name="1.0.2"></a>
## [1.0.2](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.0.1...v1.0.2) (2018-10-07)


### Features

* Add type definitions.



<a name="1.0.1"></a>
## [1.0.1](https://github.com/vivaxy/react-native-auto-height-image/compare/v1.0.0...v1.0.1) (2018-07-13)


### Features

* Reformat codes.



<a name="1.0.0"></a>

ExampleApp/App.js
import React, { Component } from 'react';
import AutoHeightImage from 'react-native-auto-height-image';
import {
  StyleSheet,
  Text,
  ScrollView,
  TextInput,
  Animated
} from 'react-native';

import image from './assets/image.png';

export default class App extends Component {
  state = {
    dynamicWidth: 200,
    fadeAnim: new Animated.Value(0)
  };

  handleTextInputChange = (text) => {
    const width = Number(text);
    if (!Number.isNaN(width)) {
      this.setState({ dynamicWidth: width });
    }
  };

  fadeIn = () => {
    Animated.timing(this.state.fadeAnim, {
      toValue: 1,
      duration: 5000,
      useNativeDriver: true
    }).start();
  };

  componentDidMount() {
    this.fadeIn();
  }

  render() {
    const { dynamicWidth } = this.state;
    return (
      <ScrollView
        style={styles.scrollViewContainer}
        contentContainerStyle={styles.scrollViewContentContainer}
      >
        <TextInput
          value={String(dynamicWidth)}
          keyboardType="numeric"
          style={styles.textInputStyle}
          onChangeText={this.handleTextInputChange}
        />
        <Text>Basic example</Text>
        <AutoHeightImage
          width={100}
          source={{ uri: 'http://placehold.it/350x150' }}
        />
        <Text>Basic example with local image</Text>
        <AutoHeightImage width={100} source={image} />
        <Text>Basic example with dynamic width</Text>
        <AutoHeightImage
          width={dynamicWidth}
          maxHeight={300}
          source={{ uri: 'http://placehold.it/350x150' }}
        />
        <Text>Basic example with dynamic width and local image</Text>
        <AutoHeightImage width={dynamicWidth} source={image} />
        <Text>Wrong image</Text>
        <AutoHeightImage
          width={100}
          source={{ uri: 'https://vivaxy.github.io/404' }}
          onError={(error) => {
            console.log('----- onError', error);
          }}
        />
        <Text>Wrong image with fallback</Text>
        <AutoHeightImage
          width={100}
          source={{ uri: 'https://vivaxy.github.io/404' }}
          fallbackSource={{ uri: 'http://placehold.it/350x150' }}
          onError={(error) => {
            console.log('----- onError', error);
          }}
        />
        <Text>Wrong image with local fallback</Text>
        <AutoHeightImage
          width={100}
          source={{ uri: 'https://vivaxy.github.io/404' }}
          fallbackSource={image}
          onError={(error) => {
            console.log('----- onError', error);
          }}
        />
        <Text>AnimatableImage</Text>
        <AutoHeightImage
          width={dynamicWidth}
          source={{
            uri:
              'https://upload.wikimedia.org/wikipedia/commons/1/18/React_Native_Logo.png'
          }}
          animated={true}
          style={[
            styles.fadingContainer,
            {
              opacity: this.state.fadeAnim
            }
          ]}
        />
        <Text>ImageBackground</Text>
        <AutoHeightImage
          width={dynamicWidth}
          source={{
            uri:
              'https://upload.wikimedia.org/wikipedia/commons/1/18/React_Native_Logo.png'
          }}
        >
          <Text style={styles.textStyle}>
            You can make any Child Component!
          </Text>
        </AutoHeightImage>
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  scrollViewContainer: {
    flex: 1,
    backgroundColor: '#fff',
    marginTop: 20
  },
  scrollViewContentContainer: {
    alignItems: 'center',
    paddingTop: 100
  },
  textInputStyle: {
    width: 300,
    height: 30,
    borderStyle: 'solid',
    borderColor: '#eee',
    borderWidth: 1
  },
  fadingContainer: {
    paddingVertical: 8,
    paddingHorizontal: 16,
    backgroundColor: 'powderblue'
  },
  textStyle: {
    color: 'white'
  }
});
